### PR TITLE
feat: superuser remove applications (FPLA-2405)

### PR DESCRIPTION
**Pull request by joez-f**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-2405

### Change description ###

Extend remove orders event to allow removal of applications

does NOT support removal of legacy C2s that were uploaded before the new upload additional applications event

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RemoveOrderController.java [hunk 1]
```diff
@@ -35,8 +38,9 @@
 @RequiredArgsConstructor(onConstructor = @__(@Autowired))
 public class RemoveOrderController extends CallbackController {
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RemoveOrderController.java:39]: maybe RemoveOrdersAndApplicationsController? or just RemovalController

File: ccd-definition/AuthorisationCaseField/CareSupervision/superuser.json [hunk 2]
```diff
@@ -35,13 +63,40 @@
...
+    "UserRole": "caseworker-publiclaw-superuser",
+    "CRUD": "CRU"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "c2ApplicationToBeRemoved",
+    "UserRole": "caseworker-publiclaw-superuser",
+    "CRUD": "CRU"
+  },  {
```
Comment (joez-f) [inline ccd-definition/AuthorisationCaseField/CareSupervision/superuser.json:79]: spacing json fix

File: ccd-definition/CaseEventToFields/CareSupervision/removeOrder.json [hunk 3]
```diff
@@ -48,7 +74,21 @@
...
     "PageColumnNumber": 1,
-    "ShowSummaryChangeOption": "N"
+    "ShowSummaryChangeOption": "N",
+    "FieldShowCondition": "removeOrderOrApplication=\"ORDER\""
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "removeOrder",
+    "CaseFieldID": "removeApplicationWarning",
```
Comment (joez-f) [inline ccd-definition/CaseEventToFields/CareSupervision/removeOrder.json:84]: missing permission?

File: ccd-definition/CaseEventToFields/CareSupervision/removeOrder.json [hunk 4]
```diff
@@ -98,27 +152,67 @@
...
     "PageDisplayOrder": 2,
     "PageColumnNumber": 1,
     "ShowSummaryChangeOption": "N",
+    "FieldShowCondition": "removeOrderOrApplication=\"APPLICATION\""
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "removeOrder",
+    "CaseFieldID": "otherApplicationToBeRemoved",
```
Comment (joez-f) [inline ccd-definition/CaseEventToFields/CareSupervision/removeOrder.json:188]: might need show condition to handle white space?

File: ccd-definition/AuthorisationCaseField/CareSupervision/superuser.json [hunk 5]
```diff
@@ -35,13 +63,40 @@
     "UserRole": "caseworker-publiclaw-superuser",
     "CRUD": "CRU"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "applicationToBeRemovedLabel",
+    "UserRole": "caseworker-publiclaw-superuser",
+    "CRUD": "CRU"
```
Comment (joez-f) [inline ccd-definition/AuthorisationCaseField/CareSupervision/superuser.json:71]: R

File: ccd-definition/CaseField/CareSupervision/removeOrder.json [hunk 6]
```diff
@@ -45,6 +92,33 @@
     "SecurityClassification": "Public",
     "Searchable": "N"
   },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "ID": "applicationToBeRemovedLabel",
```
Comment (joez-f) [inline ccd-definition/CaseField/CareSupervision/removeOrder.json:98]: might be redundant if show conditions on the two document ifelds

File: e2e/pages/events/removeOrderEvent.page.js [hunk 7]
```diff
@@ -2,17 +2,37 @@ const {I} = inject();
...
+    I.waitForElement(this.fields.orderList);
     await I.runAccessibilityTest();
-    I.selectOption(this.fields.list, option);
+    I.selectOption(this.fields.orderList, option);
+  },
+
+  async selectApplicationToRemove(option) {
+    I.click(this.fields.selection.application);
+    I.waitForElement(this.fields.applicationList);
+    await I.runAccessibilityTest();
```
Comment (joez-f) [inline e2e/pages/events/removeOrderEvent.page.js:26]: move accessibility test to test file

File: e2e/tests/hmctsSuperUserRemovesOrders_test.js [hunk 8]
```diff
@@ -112,14 +113,31 @@ Scenario('HMCTS super user removes a draft order from a case', async ({I, caseVi
...
+  const labelToSelect = 'C2, ' + applicationToRemove.uploadedDateTime;
+
+  await caseViewPage.goToNewActions(config.superUserActions.removeOrdersAndApplications);
+  await removeOrderEventPage.selectApplicationToRemove(labelToSelect);
+  await I.goToNextPage();
+
+  await removeOrderEventPage.addRemoveApplicationReason('Mistakes were made');
+  await I.completeEvent('Submit');
+
+  caseViewPage.checkTabIsNotPresent(caseViewPage.tabs.otherApplications);
```
Comment (joez-f) [inline e2e/tests/hmctsSuperUserRemovesOrders_test.js:130]: lets store the removed applications on the case, only available to super user? maybe ask business

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/RemoveOrderControllerAboutToSubmitTest.java [hunk 9]
```diff
@@ -85,28 +93,34 @@ void shouldUpdateGeneratedOrderCollectionAndHiddenGeneratedOrderCollection() {
...
         AboutToStartOrSubmitCallbackResponse response = postAboutToSubmitEvent(caseDetails);
 
         assertThat(response.getData()).doesNotContainKeys(
             "removableOrderList",
             "reasonToRemoveOrder",
             "orderToBeRemoved",
+            "c2ApplicationToBeRemoved",
+            "otherApplicationToBeRemoved",
             "orderTitleToBeRemoved",
+            "applicationTypeToBeRemoved",
```
Comment (joez-f) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/RemoveOrderControllerAboutToSubmitTest.java:123]: check we're not missing any fields here

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RemoveOrderController.java [hunk 10]
```diff
@@ -35,8 +38,9 @@
...
     private static final String REMOVABLE_ORDER_LIST_KEY = "removableOrderList";
+    private static final String REMOVABLE_APPLICATION_LIST_KEY = "removableApplicationList";
     private final ObjectMapper mapper;
-    private final RemoveOrderService service;
+    private final RemovalService service;
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RemoveOrderController.java:43]: maybe RemoveOrdersAndApplicationsService?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RemoveOrderController.java [hunk 11]
```diff
@@ -57,15 +62,26 @@ public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody Callback
...
+            UUID removedApplicationId = getDynamicListSelectedValue(caseData.getRemovableApplicationList(), mapper);
+            AdditionalApplicationsBundle application = service.getRemovedApplicationById(
+                caseData, removedApplicationId).getValue();
 
-        service.populateSelectedOrderFields(caseData, caseDetailsMap, removedOrderId, removableOrder);
+            service.populateApplicationFields(caseDetailsMap, application);
 
-        // Can be removed once dynamic lists are fixed
-        caseDetailsMap.put(REMOVABLE_ORDER_LIST_KEY, service.buildDynamicListOfOrders(caseData, removedOrderId));
+            // Can be removed once dynamic lists are fixed
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/RemoveOrderController.java:72]: does this actually happen - check maybe

File: service/src/main/java/uk/gov/hmcts/reform/fpl/exceptions/removeorder/RemovableOrderNotFoundException.java [hunk 12]
```diff
@@ -8,6 +8,6 @@ public RemovableOrderNotFoundException() {
     }
 
     public RemovableOrderNotFoundException(UUID orderId) {
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/exceptions/removeorder/RemovableOrderNotFoundException.java:10]: maybe name of exception should reflect it handles applications as well

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/AdditionalApplicationsBundle.java [hunk 13]
```diff
@@ -16,4 +19,15 @@
...
+
+    @JsonIgnore
+    public String toLabel() {
+        if (isNotEmpty(c2DocumentBundle) && isNotEmpty(otherApplicationsBundle)) {
+            return String.format("C2, %s", otherApplicationsBundle.toLabel());
+        } else if (isNotEmpty(c2DocumentBundle)) {
+            return c2DocumentBundle.toLabel();
+        }
+        return otherApplicationsBundle.toLabel();
+
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/AdditionalApplicationsBundle.java:31]: white space

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/AdditionalApplicationsBundle.java [hunk 14]
```diff
@@ -16,4 +19,15 @@
...
     private PBAPayment pbaPayment;
+
+    @JsonIgnore
+    public String toLabel() {
+        if (isNotEmpty(c2DocumentBundle) && isNotEmpty(otherApplicationsBundle)) {
+            return String.format("C2, %s", otherApplicationsBundle.toLabel());
+        } else if (isNotEmpty(c2DocumentBundle)) {
+            return c2DocumentBundle.toLabel();
+        }
+        return otherApplicationsBundle.toLabel();
```
Comment (joez-f) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/common/AdditionalApplicationsBundle.java:30]: might be NPE watch out!

File: service/src/main/java/uk/gov/hmcts/reform/fpl/exceptions/removeorder/MissingApplicationException.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.hmcts.reform.fpl.exceptions.removeorder;
+
+public class MissingApplicationException extends IllegalStateException {
+    public MissingApplicationException() {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/exceptions/removeorder/MissingApplicationException.java:4]: not used, can be removed

File: service/src/main/java/uk/gov/hmcts/reform/fpl/enums/OrderOrApplication.java [hunk 16]
```diff
@@ -0,0 +1,5 @@
+package uk.gov.hmcts.reform.fpl.enums;
+
+public enum OrderOrApplication {
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/enums/OrderOrApplication.java:3]: `RemovableType`?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java [hunk 17]
```diff
@@ -438,8 +444,10 @@ public String getOrderAppliesToAllChildren() {
     }
 
     private final Object removableOrderList;
+    private final Object removableApplicationList;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java:447]: Can just be DynamicList, they seem to have fixed the parsing issue

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/removeorder/RemoveApplicationServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+import uk.gov.hmcts.reform.fpl.model.order.generated.GeneratedOrder;
+import uk.gov.hmcts.reform.fpl.utils.CaseDetailsMap;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import javax.annotation.processing.Generated;
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/removeorder/RemoveApplicationServiceTest.java:23]: not used, remove

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/removeorder/RemoveApplicationServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static uk.gov.hmcts.reform.fpl.enums.OtherApplicationType.C17_EDUCATION_SUPERVISION_ORDER;
+import static uk.gov.hmcts.reform.fpl.enums.OtherApplicationType.C1_APPOINTMENT_OF_A_GUARDIAN;
+import static uk.gov.hmcts.reform.fpl.enums.OtherApplicationType.C4_WHEREABOUTS_OF_A_MISSING_CHILD;
+import static uk.gov.hmcts.reform.fpl.utils.CaseDetailsMap.caseDetailsMap;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.element;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.wrapElements;
```
Comment (lewisbirks) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/removeorder/RemoveApplicationServiceTest.java:32]: not used remove

File: service/src/main/java/uk/gov/hmcts/reform/fpl/exceptions/removeorder/MissingApplicationException.java [hunk 20]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.hmcts.reform.fpl.exceptions.removeorder;
```
Comment (lewisbirks) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/exceptions/removeorder/MissingApplicationException.java:1]: maybe rename package

**Review by aliveni — APPROVED** [review]

**Pull request merged**